src/environmentd/src/http/sql.rs

```diff
@@ -53,11 +56,75 @@ struct ErrorResponse {
     error: String,
 }
 
-pub async fn handle_sql_ws(client: AuthedClient, ws: WebSocketUpgrade) -> impl IntoResponse {
-    ws.on_upgrade(|ws| async move { run_ws(client, ws).await })
+pub async fn handle_sql_ws(
+    State(state): State<WsState>,
+    ws: WebSocketUpgrade,
+) -> impl IntoResponse {
+    ws.on_upgrade(|ws| async move { run_ws(&state, ws).await })
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+//#[serde(untagged)]
+pub struct WebSocketAuth {
+    pub user_id: String,
+    pub password: String,
+}
+
+async fn init_ws(
+    WsState {
+        frontegg,
+        adapter_client,
+    }: &WsState,
+    ws: &mut WebSocket,
+) -> Result<AuthedClient, anyhow::Error> {
+    let init_msg = ws.recv().await.ok_or_else(|| anyhow::anyhow!("closed"))??;
+    let ws_auth: WebSocketAuth = loop {
+        match init_msg {
+            Message::Text(data) => break serde_json::from_str(&data)?,
+            Message::Binary(data) => break serde_json::from_slice(&data)?,
+            // Handled automatically by the server.
+            Message::Ping(_) => {
+                continue;
+            }
+            Message::Pong(_) => {
+                continue;
+            }
+            Message::Close(_) => {
+                anyhow::bail!("closed");
+            }
+        }
+    };
+    let basic = Authorization::basic(&ws_auth.user_id, &ws_auth.password);
+    let mut headers = HeaderMap::new();
+    headers.typed_insert(basic);
+    let user = auth(frontegg, None, None, &headers).await?;
+    AuthedClient::new(adapter_client, user).await.err_into()
 }
 
-async fn run_ws(mut client: AuthedClient, mut ws: WebSocket) {
+async fn run_ws(state: &WsState, mut ws: WebSocket) {
+    let mut client = match init_ws(state, &mut ws).await {
+        Ok(client) => client,
+        Err(e) => {
+            let _ = ws
+                .send(Message::Close(Some(CloseFrame {
+                    code: 1002, // Protocol close code
```

Comment:
Can we use https://docs.rs/tungstenite/latest/tungstenite/protocol/frame/coding/enum.CloseCode.html#variant.Protocol instead? `CloseCode::Protocol.into()`